Python tracing docs refresh (#13125)
getsentry/sentry-docs: docs/platforms/python/tracing/span-lifecycle/index.mdx

---

> title: Span Lifecycles

Reviewer: ```suggestion
title: Span Lifecycle
```

description: "Learn how to add attributes to spans in Sentry to monitor performance and debug applications "
sidebar_order: 40
---

<Alert>

To capture transactions and spans customized to your organization's needs, you must first <PlatformLink to="/tracing/">set up tracing.</PlatformLink>

</Alert>

> To add custom performance data to your application, you need to add custom instrumentation in the form of [spans](/concepts/key-terms/tracing/distributed-tracing/#traces-transactions-and-spans). Spans are a way to measure the time it takes for a specific action to occur. For example, you can create a span to measure the time it takes for a function to execute.

Reviewer: ```suggestion
To add custom attributes or performance data to your applications traces, you need to attach these attributes to the traces [spans](/concepts/key-terms/tracing/distributed-tracing/#traces-transactions-and-spans). Spans represent the individual actions that occur within an overall application interaction. This might include items loading, database calls, API returns, or more.  
```

This is technically inaccurate - Spans are actually the individual page actions within an existing trace. They do have a time component to them in the sense that they measure how long a thing happens; but the semantics here are that theres a lot of things that happen inside a span that we might care about (including the outputs of what a span might "do")

## Span Lifecycle

In Python, spans are typically created using a context manager, which automatically manages the span's lifecycle. When you create a span using a context manager, the span automatically starts when entering the context and ends when exiting it. This is the recommended approach for most scenarios.

```python
import sentry_sdk

# Start a span for a task
with sentry_sdk.start_span(op="task", name="Create User"):
    # Your code here
    # The span will automatically end when exiting this block
    user = create_user(email="user@example.com")
    send_welcome_email(user)
    # The span automatically ends here when the 'with' block exits
```

Alternatively, you can create spans manually and control their lifecycle yourself, which gives you more flexibility but also more responsibility:

```python
import sentry_sdk

span = sentry_sdk.start_span(op="task", name="Create User")
try:
    # Your code here
finally:
    # Remember to always end your spans
    # If you don't call span.finish(), the span will remain open indefinitely,
    # causing memory leaks and incorrect timing data
    span.finish()

> ```

Author: Is there any common use case in which this would unblock a developer in setting up tracing?

> ```

Reviewer 2: FWIW I would even say it's likely okay to omit the note about __enter__ and __exit__

| `description` | `string`        | A longer description of what the span is doing. |
| `start_time`  | `datetime/float`| The start time of the span.                     |

## Using the Context Manager

For most scenarios, we recommend using the context manager approach with `sentry_sdk.start_span()`. This creates a new span that automatically starts when entering the context and ends when exiting it.

```python
import sentry_sdk

with sentry_sdk.start_span(op="db", name="Query Users") as span:
    # Perform a database query
    users = db.query("SELECT * FROM users")
    
    # You can set attributes on the span
    span.set_attribute("user_count", len(users))
```

The context manager also correctly handles exceptions, marking the span as failed if an exception occurs:

```python
import sentry_sdk

try:
    with sentry_sdk.start_span(op="http", name="Call External API"):
        # If this raises an exception, the span will be marked as failed
        response = requests.get("https://api.example.com/data")
        response.raise_for_status()
except Exception:
    # The span is already marked as failed and has ended
    pass
```

> ## Creating Spans Manually
> 
> In some cases, you might need more control over when a span starts and ends. For this, you can create spans manually:
> 
> ```python
> import sentry_sdk
> 
> def process_file(file_path):
>     # Create a span
>     span = sentry_sdk.start_span(op="file", name="Process File")
>     
>     try:
>         # Your processing logic
>         with open(file_path, 'r') as f:
>             content = f.read()
>         
>         # Set additional attributes
>         span.set_attribute("file_size", len(content))
>         
>         # Process the content
>         result = process_content(content)
>         
>         return result
>     except Exception as e:
>         # Mark the span as failed
>         span.set_status("error")
>         # You can also add error details as attributes
>         span.set_attribute("error_type", type(e).__name__)
>         span.set_attribute("error_message", str(e))
>         raise
>     finally:
>         # Always end the span
>         span.finish()
> ```

Reviewer 2: Following from my previous suggestion, I would delete this section on manual span management.

```suggestion
```

The example is also not very good, since (as far as I can tell), it would be possible and in fact easier to just use the `with sentry_sdk.start_span` syntax here, like so:

```python
with sentry_sdk.start_span(...) as span:  # note the "as span"
    try:
         ...
    except Exception as e:
         span.set_status("error")
         span.set_attribute(...)
         ...
    # no need for finally
```

Author: @codyde - did you have any particular users in mind when thinking about this use case?

## Getting the Current Span

You can access the currently active span using `sentry_sdk.get_current_span()`:

```python
import sentry_sdk

# Get the current active span
current_span = sentry_sdk.get_current_span()
if current_span:
    current_span.set_attribute("key", "value")
```

## Working with Transactions

[Transactions](/product/insights/overview/transaction-summary/#what-is-a-transaction) are a special type of span that represent a complete operation in your application, such as a web request. You can create transactions explicitly:

```python
import sentry_sdk

with sentry_sdk.start_transaction(name="Background Task", op="task") as transaction:
    # Your code here
    
    # You can add child spans to the transaction
    with sentry_sdk.start_span(op="subtask", name="Data Processing"):
        # Process data
        pass
```

> ## Distributed Tracing

Author: Fair, I'll take it out.
